src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1728 @@
+
+        discrete_time_step = nn.functional.softplus(discrete_time_step)
+
+        # 3.b. Discretization: B and C to [batch, seq_len, intermediate_size, ssm_state_size] (SRAM)
+        A = -torch.exp(self.A_log.float())
+        discrete_A = torch.exp(A[:, None, :, None, :] * discrete_time_step[:, :, :, :, None])  # [n_mamba_heads, batch, intermediate_size / n_mamba_heads, seq_len, ssm_state_size]
+        discrete_B = discrete_time_step[:, :, :, :, None] * B[:, :, None, :, :].float()  # [n_mamba_heads, batch, intermediate_size / n_mamba_heads, seq_len, ssm_state_size]
+        deltaB_u = discrete_B * hidden_states[:, :, :, :, None].float()  # [n_mamba_heads, batch, intermediate_size / n_mamba_heads, seq_len, ssm_state_size]
+        # 3.c perform the recurrence y ← SSM(A, B, C)(x)
+        scan_outputs = []
+        for i in range(seq_len):
+            ssm_state = discrete_A[:, :, :, i, :].transpose(0, 1) * ssm_state + deltaB_u[:, :, :, i, :].transpose(
+                0, 1
+            )  # [batch, n_mamba_heads, intermediate_size / n_mamba_heads, ssm_state_size]
+            scan_output = torch.matmul(ssm_state.transpose(0, 1).to(dtype), C[:, :, i, :].unsqueeze(-1))  # [n_mamba_heads, batch, intermediate_size / n_mamba_heads, 1]
+            scan_outputs.append(scan_output[:, :, :, 0])
+        scan_output = torch.stack(scan_outputs, dim=-1)  # [n_mamba_heads, batch, intermediate_size / n_mamba_heads, seq_len]
+        scan_output = scan_output + (hidden_states * self.D[:, None, :, None])  # [n_mamba_heads, batch, intermediate_size / n_mamba_heads, seq_len]
+        scan_output = scan_output * self.act(gate)
+
+        if use_cache:
+            cache_params.ssm_states[self.layer_idx] = ssm_state
+
+        # 4. Final linear projection
+        contextualized_states = self.out_proj(
+            scan_output.transpose(0, 1).reshape(batch_size, -1, seq_len).transpose(1, 2)
+        )  # [batch, seq_len, hidden_size]
+        return contextualized_states
+
+    def forward(self, hidden_states, cache_params: HybridMambaAttentionDynamicCache = None, attention_mask=None):
+        if self.use_fast_kernels:
+            if not is_fast_path_available or "cuda" not in self.x_proj_weight.device.type:
+                raise ValueError(
+                    "Fast Mamba kernels are not available. Make sure to they are installed and that "
+                    "the mamba module is on a CUDA device. lease run 'pip install causal-conv1d>=1.2.0' "
+                    "and 'pip install mamba-ssm', or set use_mamba_kernels=False in the model's config."
+                )
+            return self.cuda_kernels_forward(hidden_states, cache_params, attention_mask=attention_mask)
+        return self.slow_forward(hidden_states, cache_params, attention_mask=attention_mask)
+# fmt: on
+
+
+class ZambaMLP(nn.Module):
+    """
+    Adapted from transformers.models.gemma.modeling_gemma.GemmaMLP, with the flipped convention:
+    `up_proj` <-> `gate_proj`.
+    """
```

Comment:
what' s the motivation behind this? 😅 I won't push as it would be stupid to have a conversion script just for that but it' s a bit of a pity

Reply:
it was an overlook coming from converting from an internal codebase nomenclature